server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java

```diff
@@ -50,63 +48,115 @@ public TransferManager(final BlobContainer blobContainer, final FileCache fileCa
-                if (cachedIndexInput == null) {
-                    try {
-                        return new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), downloadBlockLocally(blobFetchRequest));
-                    } catch (IOException e) {
-                        logger.warn("Failed to download " + blobFetchRequest.getFilePath(), e);
-                        return null;
+        return AccessController.doPrivileged((PrivilegedAction<FileCachedIndexInput>) () -> {
+            try {
+                if (Files.exists(request.getFilePath()) == false) {
+                    try (
+                        InputStream snapshotFileInputStream = blobContainer.readBlob(
+                            request.getBlobName(),
+                            request.getPosition(),
+                            request.getLength()
+                        );
+                        OutputStream fileOutputStream = Files.newOutputStream(request.getFilePath());
+                        OutputStream localFileOutputStream = new BufferedOutputStream(fileOutputStream)
+                    ) {
+                        snapshotFileInputStream.transferTo(localFileOutputStream);
                     }
-                } else {
-                    if (cachedIndexInput.isClosed()) {
-                        // if it's already in the file cache, but closed, open it and replace the original one
-                        try {
-                            final IndexInput luceneIndexInput = blobFetchRequest.getDirectory()
-                                .openInput(blobFetchRequest.getFileName(), IOContext.READ);
-                            return new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), luceneIndexInput);
-                        } catch (IOException e) {
-                            logger.warn("Failed to open existing file for " + blobFetchRequest.getFilePath(), e);
-                            return null;
-                        }
-                    }
-                    // already in the cache and ready to be used (open)
-                    return cachedIndexInput;
                 }
-            })
-        );
+                final IndexInput luceneIndexInput = request.getDirectory().openInput(request.getFileName(), IOContext.READ);
+                return new FileCachedIndexInput(fileCache, request.getFilePath(), luceneIndexInput);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    /**
+     * Implementation of CachedIndexInput the defers creation of the underlying
+     * IndexInput until the first invocation of {@link #getIndexInput()}. This
+     * class is thread safe, and concurrent calls to {@link #getIndexInput()} will
+     * result in blocking until the initial thread completes the creation of the
+     * IndexInput.
+     */
+    private static class DelayedCreationCachedIndexInput implements CachedIndexInput {
```

Comment:
This implementation is pretty simple and just uses intrinsic locks to give thread safety to getIndexInput() and close(). I think this is okay because we don't need the locks for the operation given to `cache.compute()`. I did implement an alternative here that uses a ConcurrentHashMap along with a CompletableFuture that (I think) avoids some of the locking, particularly on the getIndexInput() case after the file has been downloaded: https://gist.github.com/andrross/1727dce13fa962404c54f527fa5f2e0f